Add a property modifier for the UI integration.

wt opened the pull request:
I have a tool script class that has a property that needs to update
the config warning when it's value changes. I thought this seemed
sufficiently generic that it should exist in the wild. :)

BTW, I looked for an appropriate place to document this, but I didn't find any place where the NativeClass derive macro was documented. I would really like to do that. I also wasn't sure where to put tests. The current test that are there don't fail, so there is that.

Here's an example of how it works on an example property:
```
#[derive(gdnative::NativeClass)]
#[inherit(Node)]
pub struct NewNode {
    #[property(update_config_warning_on_set)]
    invalid_when_false: bool,
}

#[gdnative_derive::methods]
impl NewNode {
    fn new(_owner: &Node) -> Self {
        Self {
            invalid_when_false: false,
        }
    }

    #[export]
    fn _get_configuration_warning (&self, owner: &Node) -> GodotString {
        match self.invalid_when_false {
            false => "invalid_when_false is false... tsk tsk",
            true => "",
        }
    }
}
```

If you add that as a tool class and then add a node with that script, you will get a warning that shows up and disappears at the right times in the godot editor gui. In my personal project, I went ahead and made a EditorPlugin that registers my classes so I can pick it from the gui new node type selector.

This should probably be improved so somehow only allow this on Node derived NativeClasses. Again, not sure how to do that. Adding it to a non-Node property will cause a compile-time error, so the user won't be able to get away with it. They just won't get a great error message.

wt commented:
That last update was a cargo fmt run.

toasteater commented on gdnative-derive/src/native_script.rs:
```diff
@@ -36,12 +36,20 @@ pub(crate) fn derive_native_class(input: TokenStream) -> TokenStream {
                     .with_ref_getter(|this: &#name, _| &this.#ident)
-                    .with_setter(|this: &mut #name, _, v| this.#ident = v)
+                    .with_setter(|this: &mut #name, _owner: ::gdnative::TRef<Self::Base>, v| {
+                        #update_config_warning_on_set_snippet
+                        this.#ident = v
```
Perhaps the function should be called after the value is set? Technically Godot may decide to update it immediately after `update_configuration_warning` is called in a future version.

wt replied:
Good point. Will update.

toasteater commented on gdnative-derive/src/native_script.rs:
```diff
@@ -36,12 +36,20 @@ pub(crate) fn derive_native_class(input: TokenStream) -> TokenStream {
                 None
             };
 
+            let update_config_warning_on_set_snippet = match config.update_config_warning_on_set {
+                true => quote!(_owner.update_configuration_warning();),
+                false => quote!(),
+            };
```
Shouldn't this just be an `if` expression...?

wt replied:
I like the way match forces handling all possibilities and allows just setting the variable appropriately.

toasteater replied:
`if` is an expression.

```rust
let update_config_warning_on_set_snippet = if config.update_config_warning_on_set {
    quote!(_owner.update_configuration_warning();)
} else {
    quote!()
};
```

toasteater reviewed (commented):
Thanks for the PR! While I agree that the ability to modify the behavior of getters and setters generated by `#[property]` is great to have, I wonder if the solution in the current PR is way too specific. If we add a generic way to add hooks instead, both the problems with non-`Node` base classes and testability should resolve themselves. For example:

```rust
#[derive(NativeScript)]
struct Foo {
    // this will call `utils::update_configuration_warning` with `&self` and `owner` after the field is set.
    #[property(after_set = "utils::update_configuration_warning")]
    bar: bool,
}

/// this will be a wrapper in user code, that can be shared between multiple scripts
mod utils {
    pub fn update_configuration_warning<C>(_script: &mut C, owner: TRef<C::Base, Shared>)
    where
        C: NativeClass,
        C::Base: SubClass<Node>,
    {
        owner.upcast().update_configuration_warning();
    }
}
```

...then, for tests, we can just make it modify some field in the script struct. Since the user provides the hook in question, they'll also get proper errors attributed to the functions they provide when the types mismatch.

toasteater commented:
I think doc-comments on proc-macros can be put in the `gdnative-derive` crate on the respective top-level function.

wt commented:
> Thanks for the PR!

YW

> While I agree that the ability to modify the behavior of getters and setters generated by `#[property]` is great to have, I wonder if the solution in the current PR is way too specific. If we add a generic way to add hooks instead, both the problems with non-`Node` base classes.

I hear you on this, and I had a more generic implementation before. However, I think this would be a very common case for tool script classes. That's why I ultimately implemented this.

I'm not sure what adding the util function in your suggestion does. Mine already will complain about types not working if use my modifier on a non-Node. You solution further required Node-based object properly and gave a better error message. I fixed mine for this. In your solution, I am still struggling with this. How would we implement multiple util functions that may have some ordering requirements. My attempt was to make this more declarative, which kinda landed me here.

FWIW, I do think your suggestion on allowing a function does make sense. I am just not sure if it should be the only option.

>  If we add a generic way to add hooks instead, both the problems with non-Node base classes and testability should resolve themselves. For example:

I don't necessarily buy that this would fix the non-Node base issue. However, your solution had a nice property of doing the Node subclass detection, so I added that to mine. Here's an example error when applying my modifier to an attr in a Resource-based struct:

```
error[E0277]: the trait bound `gdnative::gdnative_bindings::Resource: gdnative::prelude::SubClass<gdnative::prelude::Node>` is not satisfied
  --> src/widget.rs:10:10
   |
10 | #[derive(gdnative::NativeClass)]
   |          ^^^^^^^^^^^^^^^^^^^^^ the trait `gdnative::prelude::SubClass<gdnative::prelude::Node>` is not implemented for `gdnative::gdnative_bindings::Resource`
   |
   = help: the following implementations were found:
             <gdnative::gdnative_bindings::Resource as gdnative::prelude::SubClass<gdnative::prelude::Object>>
             <gdnative::gdnative_bindings::Resource as gdnative::prelude::SubClass<gdnative::prelude::Reference>>
   = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

I think that's probably as good as your solution would do.

My main problem with the above implementation is that it shows the wrong line numbers/offsets for the issue. I would like to find a way to fix that.

FWIW, my views are not super strongly opposed to any of you ideas, and I am open to changing the code.

toasteater commented:
> I'm not sure what adding the util function in your suggestion does.

It's a demonstration of a provisional API? If you only allow calling associated methods, then users may have to copy-paste the same methods across multiple script types. Allowing users to put accessor hooks into outside modules solves that. It's what the user asks to be called after the property is set and could literally be anything. There is no inherent meaning in what it's called and what it does. I just wrote it that way to show that your original use case is covered.

> How would we implement multiple util functions that may have some ordering requirements.

Isn't that easy? Choose between:

1. Only allow one `after_set` callback. If the user needs to do multiple things, they'll make a function that does those things in whatever order they like.
2. Or, remember the order the callbacks are given in the attribute. i.e. `#[property(after_set="a", after_set="b", after_set="c")]` calls `a`, `b`, and then `c` in that order.

Personally I prefer the first option because that's simpler, but I'm not against implementing 2 if you prefer and would implement it.

> I don't necessarily buy that this would fix the non-Node base issue.

Why not? I don't see any reason why the base type would still be an issue, because the feature is generalized to all `NativeClass` types regardless of what their base classes are. You would be able to use the hypothetical `after_set` on a type extending `Reference` just fine.

> My main problem with the above implementation is that it shows the wrong line numbers/offsets for the issue. I would like to find a way to fix that.

You might be able to span the code at the `#[property]` attribute instead of the call site (the `#[derive]` attribute in this case) which is used by default.

toasteater commented:
> I think this would be a very common case for tool script classes. That's why I ultimately implemented this.

But in the end `update_config_warning_on_set` is just too ad-hoc. What if you want to also do something else after a property is set? That's what setters are for after all. If general setter hooks are implemented, this case can be easily covered in userland with just the inclusion of an one-liner function that is very easy to conceive. The general solution is also very close to your existing code, so I would consider it simply unreasonable to hard-code that one single case into the library. This is a Rust codebase, not a Go one.

wt commented:
What about adding a new crate/module with the util function. Then other recipes could land there.

More advanced users could create their own.

Would something like that be acceptable?

My main concern is that I want a place for more common recipes to be put so that users don't have to figure this out every time.

toasteater commented:
Yes, a crate containing utility functions is currently considered for issues like #410, and can work for this case as well. Currently the main issue with a utils crate is that it will have to depend on `gdnative-bindings`, which is generated and may have to be replaced by users if they use custom versions of Godot, making dependency a little tricky. This might be solved by introducing a few traits in the utils trait that can be automatically implemented by the bindings generator, but work needs to be done to see if that works and solves the problem.

That being said, I still don't think a version of `update_configuration_warning` that simply drops the first argument and calls the same method on the second argument is a good candidate for inclusion in the crate. It's not as painful as other things considered for inclusion, generally involving getting typed nodes from the scene tree.

toasteater commented:
If you think there should be a `update_configuration_warning` in the utils crate, you may create another issue for discussion. If you agree that generalized accessor hooks is the way to go, we should focus on getting the feature done for this PR.

wt commented:
I am focusing on getting the hooks in. Here is a question. Are both pre and post hooks useful or do we only need a post hook? My impression is that both should be useful.

wt commented:
I've opened a new PR #581 to address this. Let's continue discussion there please.
